include/ensmallen_bits/swats/swats_update.hpp

```diff
@@ -0,0 +1,176 @@
+/**
+ * @file swats_update.hpp
+ * @author Marcus Edel
+ *
+ * SWATS update rule for Switches from Adam to SGD method.
+ *
+ * ensmallen is free software; you may redistribute it and/or modify it under
+ * the terms of the 3-clause BSD license.  You should have received a copy of
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef ENSMALLEN_SWATS_SWATS_UPDATE_HPP
+#define ENSMALLEN_SWATS_SWATS_UPDATE_HPP
+
+namespace ens {
+
+/**
+ * SWATS is a simple strategy which switches from Adam to SGD when a triggering
+ * condition is satisfied. The condition relates to the projection of Adam steps
+ * on the gradient subspace.
+ *
+ * For more information, see the following.
+ *
+ * @code
+ * @article{Keskar2017,
+ *   author  = {Nitish Shirish Keskar and Richard Socher},
+ *   title   = {Improving Generalization Performance by Switching from Adam to
+ *              {SGD}},
+ *   journal = {CoRR},
+ *   volume  = {abs/1712.07628},
+ *   year    = {2017}
+ *   url     = {http://arxiv.org/abs/1712.07628},
+ * }
+ * @endcode
+ */
+class SWATSUpdate
+{
+ public:
+  /**
+   * Construct the SWATS update policy with given parameter.
+   *
+   * @param epsilon The epsilon value used to initialise the squared gradient
+   *        parameter.
+   * @param beta1 The smoothing parameter.
+   * @param beta2 The second moment coefficient.
+   */
+  SWATSUpdate(const double epsilon = 1e-8,
+              const double beta1 = 0.9,
+              const double beta2 = 0.999) :
+    epsilon(epsilon),
+    beta1(beta1),
+    beta2(beta2),
+    iteration(0),
+    phaseSGD(false),
+    sgdRate(0),
+    sgdLamda(0)
+  {
+    // Nothing to do.
+  }
+
+  /**
+   * The Initialize method is called by SGD Optimizer method before the start of
+   * the iteration update process.
+   *
+   * @param rows Number of rows in the gradient matrix.
+   * @param cols Number of columns in the gradient matrix.
+   */
+  void Initialize(const size_t rows, const size_t cols)
+  {
+    m = arma::zeros<arma::mat>(rows, cols);
+    v = arma::zeros<arma::mat>(rows, cols);
+
+    sgdV = arma::zeros<arma::mat>(rows, cols);
```

Comment:
Since `sgdV` is only used during the SGD portion of optimization, it seems like it could be possible to reuse `v` here.  I don't have a strong opinion either way though.